docs/architecture/adr-082-socket-protocol.md

```diff
@@ -0,0 +1,237 @@
+
+  Claim: If we find meaningful advantages to maintaining our own custom RPC
+  protocol in Tendermint, we should treat it as a first-class project within
+  the core and invest in making it good enough that we do not require other
+  options.
+
+**One important consideration** when discussing these questions is that _any
+outcome which includes keeping the socket protocol will have eventual migration
+impacts for out-of-process applications_ regardless. To fix the limitations of
+the socket protocol as it is currently designed will require making _breaking
+changes_ to the protocol.  So, while we may put off a migration cost for
+out-of-process applications by retaining the socket protocol in the short term,
+we will eventually have to pay those costs to fix the problems in its current
+design.
+
+## Detailed Design
+
+1. If we choose to standardize on gRPC, the main work in Tendermint core will
+   be removing and cleaning up the code for the socket client and server.
+
+   Besides the code cleanup, we will also need to clearly document a
+   deprecation schedule, and invest time in making the migration easier for
+   applications currently using the socket protocol.
+
+   > **Point for discussion:** Migrating from the socket protocol to gRPC
+   > should mostly be a plumbing change, as long as we do it during a release
+   > in which we are not making other breaking changes to ABCI. However, the
+   > effort may be more or less depending on how gRPC integration works in the
+   > application's implementation language, and would have to be sure networks
+   > have plenty of time not only to make the change but to verify that it
+   > preserves the function of the network.
+   >
+   > What questions should we be asking node operators and application
+   > developers to understand the migration costs better?
+
+2. If we choose to keep only the socket protocol, we will need to follow up
+   with a more detailed design for extending and upgrading the protocol to fix
+   the existing performance and operational issues with the protocol.
+
+   Moreover, since the gRPC interface has been around for a long time we will
+   also need a deprecation plan for it.
+
+3. If we choose to keep both options, we will still need to do all the work of
+   (2), but the gRPC implementation should not require any immediate changes.
+
+
+## Consequences
+
+- **Standardize on gRPC**
+
+   - ✅ Addresses existing performance and operational issues.
+   - ✅ Replaces custom code with a well-maintained widely-used library.
+   - ✅ Aligns with Cosmos SDK, which already uses gRPC extensively.
+   - ✅ Aligns with priv validator interface, for which the socket protocol is already deprecated for gRPC.
+   - ❓ Applications will be hard to implement in a language without gRPC support.
+   - ⛔ All users of the socket protocol have to migrate to gRPC, and we believe most current out-of-process applications use the socket protocol.
```

Comment:
One concern is that gRPC may be a bit of an overkill for local IPC:
- There's a higher latency: https://www.mpi-hd.mpg.de/personalhomes/fwerner/research/2021/09/grpc-for-ipc/
- There's more hassle for secure setups with TCP: with Unix domain sockets, one can just use file permissions... for TCP, either some firewall rules perhaps via eBPF to check only ABCI app + Tendermint are talking to each other on that port, or have some TLS certificate setup between them.

I believe gRPC can run over Unix domain sockets (or even inproc: https://github.com/grpc/grpc/tree/master/src/core/ext/transport/inproc ), but I haven't tried... but if gRPC remains as the only socket protocol, it'll be good to document or implement support for these more lightweight transport options

Reply:
gRPC does work fine over a Unix-domain socket. Ordinarily applications don't use it this way, but for ABCI applications it would be fine. An application can also terminate its listener after handshaking with the consensus node, so even with a TCP address there isn't  a practical security concern.

I don't think we (yet) have any evidence suggesting that the latency of ABCI calls (i.e., the call and response itself, ignoring time spent in the service handler) is a meaningful factor in out-of-process, and I expect permitting applications to service requests concurrently (which the socket protocol's design narrowly limits) would render a bit of additional overhead negligible. (The 100µsec discussed in the post would _already_ be negligible for most ABCI calls even before factoring in concurrency)

Do you know of anyone who's tried both options on a production Tendermint network, who might have some data about the difference?

Reply:
Another option for pipe-based communication would be to have the supervisor process (say, something like `cosmovisor`) orchestrate the startup of both processes and plumb pipe descriptors directly into the child processes. Then there's not even a rendezvous to maintain. (Though it's an interesting question whether that would matter too much, given that a lot of operators seem to run nodes in containers anyway, where it's already easy to isolate network addresses from outside access).

Reply:
> Do you know of anyone who's tried both options on a production Tendermint network, who might have some data about the difference?

I'm not sure, it may be good to reach out to Nomic or other ABCI-based projects that care about a low latency.

I tried both options in a stress-testing setup on a pre-0.33 Tendermint network a long time ago, and observed a similar 10x to >50x µsec difference as discussed in that post... which likely shouldn't matter to most ABCI calls.